cpp/src/community/k_truss_impl.cuh

```diff
@@ -0,0 +1,929 @@
+      // of the edges to be removed 'edge_triangle_count_pair_first' marks the begining of the valid
+      // edges
+      auto invalid_transposed_edge_triangle_count_first =
+        thrust::stable_partition(handle.get_thrust_policy(),
+                                 transposed_edge_triangle_count_pair_first,
+                                 transposed_edge_triangle_count_pair_first + num_triangles.size(),
+                                 [k] __device__(auto e) {
+                                   auto num_triangles = thrust::get<1>(e);
+                                   return num_triangles >= k - 2;
+                                 });
+
+      num_invalid_edges = static_cast<size_t>(
+        thrust::distance(invalid_transposed_edge_triangle_count_first,
+                         transposed_edge_triangle_count_pair_first + edgelist_srcs.size()));
+
+      if (num_invalid_edges == 0) { break; }
+
+      auto num_valid_edges = edgelist_srcs.size() - num_invalid_edges;
+
+      // case 3. For the (p, q), find intersection 'r' to create (p, r, -1) and (q, r, -1)
+      // FIXME: check if 'invalid_transposed_edge_triangle_count_first' is necessery as I operate on
+      // 'vertex_pair_buffer' which contains the ordering with the number of triangles.
+      // FIXME: debug this stage. There are edges that have been removed that are still found in nbr
+      // intersection
+      auto [intersection_offsets, intersection_indices] = detail::nbr_intersection(
+        handle,
+        cur_graph_view,
+        cugraph::edge_dummy_property_t{}.view(),
+        thrust::make_zip_iterator(edgelist_srcs.begin() + num_valid_edges,
+                                  edgelist_dsts.begin() + num_valid_edges),
+        thrust::make_zip_iterator(edgelist_srcs.end(), edgelist_dsts.end()),
+        std::array<bool, 2>{true, true},
+        do_expensive_check);
+
+      rmm::device_uvector<edge_t> num_triangles_invalid(num_invalid_edges, handle.get_stream());
+
+      // Update the number of triangles of each (p, q) edges by looking at their intersection
+      // size
+      thrust::adjacent_difference(handle.get_thrust_policy(),
+                                  intersection_offsets.begin() + 1,
+                                  intersection_offsets.end(),
+                                  num_triangles_invalid.begin());
+
+      thrust::for_each(
+        handle.get_thrust_policy(),
+        thrust::make_counting_iterator<edge_t>(0),
+        thrust::make_counting_iterator<edge_t>(num_invalid_edges),
+        unroll_edge_invalid<vertex_t, edge_t>{
+          edge_t{num_valid_edges},
+          raft::device_span<edge_t>(num_triangles.data(), num_triangles.size()),
+          raft::device_span<edge_t>(num_triangles_invalid.data(), num_triangles_invalid.size())});
+
+      size_t accumulate_pair_size = intersection_indices.size();
+
+      // FIXME: Find a way to not have to maintain a dataframe_buffer
```

Comment:
You can do this by combining the below tabulate & for_each calls (find p, r edges than run binary_search and atomic op), but it may not work in multi-GPU  as we need to shuffle p,r edges. So, let's keep this code as is for now.